psi_cardinality/cpp/psi_cardinality_client_c.cpp

```diff
@@ -0,0 +1,70 @@
+#include "psi_cardinality_client_c.h"
+
+#include "psi_cardinality_client.h"
+
+using namespace psi_cardinality;
+
+psi_cardinality_client_ctx psi_cardinality_client_create() {
+    auto result = PSICardinalityClient::Create();
+    if (result.ok()) {
+        auto value = std::move(result).ValueOrDie();
+        return value.release();
+    }
+    return nullptr;
+}
+
+void psi_cardinality_client_delete(psi_cardinality_client_ctx* ctx) {
+    auto client = static_cast<PSICardinalityClient*>(*ctx);
+    if (client == nullptr) {
+        return;
+    }
+    delete client;
+    *ctx = nullptr;
+}
+
+int psi_cardinality_client_create_request(psi_cardinality_client_ctx ctx,
+                                          client_buffer_t* inputs,
+                                          size_t input_len, char** out,
+                                          size_t* out_len) {
+    auto client = static_cast<PSICardinalityClient*>(ctx);
+    if (client == nullptr) {
+        return -1;
+    }
+    std::vector<std::string> in;
+    for (size_t idx = 0; idx < input_len; ++idx) {
+        in.push_back(std::string(inputs[idx].buff, inputs[idx].buff_len));
+    }
+
+    auto result = client->CreateRequest(in);
+    if (!result.ok()) {
+        return -1;
+    }
+    auto value = std::move(result).ValueOrDie();
+
+    size_t len = value.size() + 1;
+    *out = new char[len];
+    strncpy(*out, value.c_str(), len);
+    *out_len = len;
+
+    return 0;
+}
+
+void psi_cardinality_client_delete_buffer(psi_cardinality_client_ctx ctx,
```

Comment:
Go has its own management for the memory and threading, separate from C.

That means that when I access a C pointer, everything Go-related goes on a stack, and it passes the control to the C env/kernel.
Go never actually gets access to the C pointers. it has the pointer concept for the garbage collector, but no pointer arithmetics, for example.

TLDR: Go doesn't actually control the memory allocated in C, so it has to be freed from C. I couldn't find a prettier solution.

Reply:
I agree that you need to free the memory from C, but I thought this could be done easier by constructing the buffer using `malloc` instead of `new`, and then calling `C.free(out)`instead of `C.psi_cardinality_client_delete_buffer(c.context, &out)` [here](https://github.com/OpenMined/psi-cardinality/pull/12/files#diff-c6c24f591a29c4cd410bb99331d22b44R56).

Reply:
Ah, I understand it now. 

I thought it was better to make no assumptions about the memory in Go(malloc vs new), and to manage it only in one place.